### ISSUE TITLE
[Step8] End-to-End FP32 Inference (Image -> Softmax) & Torch Match

### ISSUE THREAD

**Issue by yeontachi** [yeontachi/DLQ#10]
# Step8: End-to-End FP32 Inference (Image → Softmax) & Torch Match

## 배경
Step2~7에서 Conv/BN/ReLU, Skip-Add, Downsample(1x1), GAP, Softmax까지 부분 경로를 검증 완료. 본 이슈에서는 **ResNet-18 FP32 전체 경로**를 하나의 실행 파일에서 수행하고, Torch 기준 결과와 `atol ≤ 1e-4`로 일치 여부 및 단계별 성능을 기록한다.

---

## 목표
- 단일 실행 파일 `build/fp32/step8_e2e`로 다음 파이프라인 수행:
  - 입력 이미지 전처리(Resize=256, CenterCrop=224, Normalize) → `conv1/bn1/relu` → **maxpool(3×3,s=2,p=1)** → `layer1~layer4` → **GAP** → **FC** → **Softmax**
- Torch 기준 **logits**/**prob**와 수치 일치(`max_abs ≤ 1e-4`) 확인
- 각 주요 구간(Stem, L1, L2, L3, L4, GAP, FC, Softmax)의 소요 시간(ms) 로그 출력
- 실패 시 어느 지점에서 오차가 커졌는지(레이어/블록 단위) 확인 가능한 로그 제공
- 한 번의 스크립트(`bash scripts/run_step8.sh`)로 **픽스처 생성 → 빌드 → 실행 → 로그 저장**까지 자동화

---

## 산출물
- 실행 파일: `build/fp32/step8_e2e`
- 스크립트: `scripts/run_step8.sh`
- 커널: `cpp/fp32/kernels/maxpool2d.cu` (신규)
- 런타임: `cpp/fp32/runtime/infer_e2e.cu` (신규)
- 픽스처: `tools/make_step8_fixture.py` (신규; Torch로 입력/정답 생성)
- CMake: `cpp/fp32/CMakeLists.txt` 타깃 추가

---

## Acceptance Criteria
- Torch 기준값과 아래 모두 충족:
  - `Diff logits max_abs ≤ 1e-4`
  - `Diff prob   max_abs ≤ 1e-4`
- 실행 로그에 아래 항목 존재:
  - `stem(maxpool incl.), l1, l2, l3, l4, gap, fc, softmax` 각 소요 시간(ms)
  - 커널 합산: `im2col, gemm, bn, relu` 시간 합계
  - 수치 차이: `Diff logits`, `Diff prob` (max_abs, mean_abs)
- 스크립트 한 번으로 **픽스처 생성 → 빌드 → 실행 → 로그 저장** 완료
- 실패 시 `[FAIL] Step8 diff exceeded`와 함께 어느 단계에서 차이가 커졌는지 메시지 포함

---

### 1) 픽스처 생성 스크립트 (신규)
`tools/make_step8_fixture.py`
- torchvision `resnet18(weights=IMAGENET1K_V1)` 로드
- 전처리(Resize256 → CenterCrop224 → ToTensor → Normalize)
- 입력 `input.bin`, 최종 `final_logits.bin`, `final_prob.bin` 저장 (`exports/resnet18/fp32/fixtures_step8/`)

### 2) MaxPool 커널 (신규)
`cpp/fp32/kernels/maxpool2d.cu`
- N=1, NCHW 입력, 일반화된 파라미터: `kH,kW,sH,sW,pH,pW`
- 기본 사용: 3×3, stride=2, pad=1

### 3) E2E 런타임 (신규)
`cpp/fp32/runtime/infer_e2e.cu`
- 유틸 함수:
  - `run_conv_bn_relu`, `run_conv_bn`, `run_add`
  - Conv는 `im2col + sgemm_tiled` 경로 사용
- 경로:
  - stem(7×7,s=2,p=3) → maxpool → layer1(64,s=1) → layer2(128,s=2) → layer3(256,s=2) → layer4(512,s=2) → GAP → FC(1000) → softmax
- 각 단계별 Timer 기록 및 누적 커널 시간(im2col/gemm/bn/relu) 집계
- Torch 기준 logits/prob와 diff 계산 및 통과 여부 출력

### 4) CMake 타깃 추가
`cpp/fp32/CMakeLists.txt`
- `add_executable(step8_e2e ...)` 추가
- `maxpool2d.cu`, `infer_e2e.cu` 포함
- `CUDA_SEPARABLE_COMPILATION ON` 설정

### 5) 실행 스크립트 (신규)
`scripts/run_step8.sh`
- 픽스처 생성 → CMake 빌드 → 실행 → 로그 저장(`logs/step8_*.log`)
- 출력:
  - `[Run] ...`
  - `[Cmd] ...`
  - `E2E done`/`[OK]`/`[FAIL]` 및 상세 수치

## 테스트 방법

```bash
# End-to-End 한 번에 수행
bash scripts/run_step8.sh

# 수동 검증(선택)
python tools/make_step8_fixture.py --manifest exports/resnet18/fp32
cmake -S cpp -B build -DCMAKE_BUILD_TYPE=Release
cmake --build build -j
build/fp32/step8_e2e --manifest exports/resnet18/fp32
````

예상 성공 로그(요지):

* `E2E done`
* `kernel sums — im2col:..., gemm:..., bn:..., relu:...`
* `Diff logits : max_abs=... mean_abs=...`
* `Diff prob   : max_abs=... mean_abs=...`
* `[OK] Step8 E2E matched within atol 1e-4`
* `Saved log -> logs/step8_YYYYmmdd_HHMMSS.log`

---

## 코드 변경 파일

* **추가**: `tools/make_step8_fixture.py`
* **추가**: `cpp/fp32/kernels/maxpool2d.cu`
* **추가**: `cpp/fp32/runtime/infer_e2e.cu`
* **수정**: `cpp/fp32/CMakeLists.txt` (step8_e2e 타깃 추가)
* **추가**: `scripts/run_step8.sh`

---

## 리스크 & 체크리스트

* [ ] manifest(.bin) 경로 불일치 → 실행 전 `--manifest` 루트 확인
* [ ] maxpool 출력 Shape(56×56) 불일치 → pad/stride/hw 계산 재검증
* [ ] downsample(1×1,s=2) 경로 weight/bn 파일명 mismatch
* [ ] FC 편향 적용 누락/차이 → host bias add 확인
* [ ] softmax 수치 안정성 → 내부에서 최대값 감산(stable softmax) 사용 여부 확인

---

## 완료 조건(체크리스트)

* [ ] `scripts/run_step8.sh` 단일 명령으로 픽스처/빌드/실행/로그 저장
* [ ] 로그에 단계별 시간 및 diff 출력
* [ ] `[OK] Step8 E2E matched within atol 1e-4` 메시지
* [ ] 코드/스크립트 리드미 반영(선택)
